Update Figma variables from JSON files (#1884)
Telefonica/mistica-design: tokens/figma/README.md

Reviewer: should we also make changes in the root README? We have a process to sync tokens with Figma tokens.. I think that is not recommended to sync tokens with Figma tokens when we use variables 🧐 wdyt?

Author: README updated

Reviewer: I think that this readme is not showing the last changes about to have two different scripts. A script for update brands, and other script to update middleware

Reviewer: maybe you can include a screenshot or the url to the tldraw schema?

Author: Added

Reviewer: I meant in this readme! jaja

# Project Overview

> This project is designed to update Figma variables based on a JSON input, primarily focused on managing brand themes, colors, and other design tokens. The project retrieves existing variables from Figma, processes the provided JSON data, and updates or creates new variables in collections "Theme" and "Skin".

Reviewer: ```suggestion
This project is designed to update Figma variables based on a JSON input, primarily focused on managing brand themes, colors, and other design tokens. The project retrieves existing variables from Figma, processes the provided JSON data, and updates or creates new variables in collections "Mode" and "Brand".
```

## Features

> - **Fetch Existing Figma Data**: Retrieves the existing variables and collections from Figma.

Reviewer 2: Why "pascal" case in here?

Author: Fixed

- **Process JSON Data**: Extracts theme and token data from provided JSON files for each brand.
- **Update or Create Variables**: Adds new variables or updates existing ones based on the brand's light and dark themes.

> - **Handle Variable Modes**: Ensures each brand's mode (e.g., "Light", "Dark") is updated or created in the Figma "Skin" collection.

Reviewer: ```suggestion
- **Handle Variable Modes**: Ensures each brand's mode (e.g., "Light", "Dark") is updated or created in the Figma "Brand" collection.
```

Author: Resolved

- **Support for Multiple Brands**: Processes multiple brands, mapping each brand's unique variables into Figma's collections.

## Setup

> ### Environment Variables:

Reviewer 2: Aren't you missing the brand variables here? (like `MOVISTAR_FILE_KEY`)

Author: Added

- `FIGMA_TOKEN`: The API token to authenticate with Figma.
- `MIDDLEWARE_KEY`: Token for file where the variables need to be created / updated

### Dependencies:

- Node.js and packages such as `node-fetch`, `dotenv`, and `fs` are used to manage API requests, read local files, and load environment variables.

## Key Functions

### `updateTheme(jsonData, brand, FILE_KEY)`

This function updates the theme variables in Figma for a specific brand. It:

- Fetches the current variables from Figma.

> - Updates modes and variables for "Light" and "Dark" themes.

Reviewer 2: Not important, but why not :P

```suggestion
- Updates modes and variables for `"Light"` and `"Dark"` themes.
```

- Sends a POST request to update Figma with the new data.

### `updateSkinColorVariables(brands, FILE_KEY)`

> This function focuses on updating color variables in the "Skin" collection. It:

Reviewer: ```suggestion
This function focuses on updating color variables in the "Brand" collection. It:
```

> - Maps color variables from the "Theme" collection to the "Skin" collection.

Reviewer: ```suggestion
- Maps color variables from the "Mode" collection to the "Brand" collection.
```

# Project overview

This project is designed to update Figma variables based on a JSON input, primarily focused on managing brand themes, colors, and other design tokens. The project retrieves existing variables from Figma, processes the provided JSON data, and updates or creates new variables in collections "Mode" and "Brand".

## Features

- **Fetch existing Figma data**: Retrieves the existing variables and collections from Figma.
- **Process JSON data**: Extracts theme and token data from provided JSON files for each brand.

> - **Update or dreate variables**: Adds new variables or updates existing ones based on the brand's light and dark themes.

Reviewer 2: ```suggestion
- **Update or create variables**: Adds new variables or updates existing ones based on the brand's light and dark themes.
```

# Design Tokens Automation Script

## Objective

This script automates the process of fetching, updating, and posting design tokens to Figma for multiple brands. It reads JSON data files, extracts relevant variables, compares them with existing data from Figma, updates them if necessary, and then posts the changes back to Figma's API.

## Overview of the Script

> ### Environment Setup

Reviewer: Local Enviroment Setup

Author: README updated

- The script uses the `dotenv` package to load environment variables (like Figma API tokens and file keys) from a `.env` file.
- It imports necessary modules (`fetch`, `fs`, `path`, etc.) and utility functions (`extractJsonData`, `hexToRgba`, `extractPaletteValue`).

### File and Data Preparation

- The script reads JSON files from a specific directory, filters out the ones with a `.json` extension, and extracts their data using the `extractJsonData` function.

### Brands and URLs

> - An object `brands` maps brand names to their corresponding Figma API URLs. This allows the script to handle multiple brands in one go.

Reviewer: is this correct?

Author: Still correct for the update-skins script, since each file should be updated
